lib/nerve.rb

```diff
@@ -15,35 +15,103 @@ class Nerve
       # required options
       log.debug 'nerve: checking for required inputs'
       %w{instance_id services}.each do |required|
-        raise ArgumentError, "you need to specify required argument #{required}" unless opts[required]
+        raise ArgumentError, "you need to specify required argument #{required}" unless config[required]
       end
-
-      @instance_id = opts['instance_id']
-      @services = opts['services']
-      @heartbeat_path = opts['heartbeat_path']
-      @watchers = {}
-
-      log.debug 'nerve: completed init'
+      @instance_id = config['instance_id']
+      @watchers_desired = config['services']
+      @heartbeat_path = config['heartbeat_path']
     end
 
     def run
       log.info 'nerve: starting run'
-
-      @services.each do |name, config|
-        launch_watcher(name, config)
-      end
-
       begin
         loop do
+          # Check if configuration needs to be reloaded and reconcile any new
+          # configuration of watchers with old configuration
+          if @config_to_load
+            load_config!
+
+            # Reap undesired service watchers
+            services_to_reap = @watchers.select{ |name, _|
+              !@watchers_desired.has_key?(name)
+            }.keys()
+            log.info "nerve: reaping old watchers: #{services_to_reap}"
+            services_to_reap.each do |name|
+                reap_watcher(name) rescue "nerve: could not cleanly reap #{name}"
+            end
+
+            # Start new desired service watchers
+            services_to_launch = @watchers_desired.select{ |name, _|
+              !@watchers.has_key?(name)
+            }.keys()
+            log.info "nerve: launching new watchers: #{services_to_launch}"
+            services_to_launch.each do |name|
+              launch_watcher(name, @watchers_desired[name])
+            end
+
+            # Detect and update existing service watchers
+            services_to_update = @watchers.select { |name, _|
+              @watchers_desired.has_key?(name)
+            }.each { |name, _|
+              new_watcher_config = merged_config(@watchers_desired[name], name)
+              if new_watcher_config.hash != @watcher_versions[name]
+                log.info "nerve: detected new config for #{name}"
+                # Keep the old watcher running until the replacement is launched
+                # This keeps the service registered while we change it over
+                # This also keeps connection pools active across diffs
+                temp_name = "#{name}_#{@watcher_versions[name]}"
+                @watchers[temp_name] = @watchers.delete(name)
+                @watcher_versions[temp_name] = @watcher_versions.delete(name)
+                log.info "nerve: launching new watcher for #{name}"
+                launch_watcher(name, @watchers_desired[name])
+                log.info "nerve: reaping old watcher #{temp_name}"
+                reap_watcher(temp_name) rescue "nerve: could not cleanly reap #{name}"
```

Comment:
for example, if someone is using the leader election code in synapse, the new ephemeral node will still cause an election as it will be the highest sequence number.

it almost seems like if we want to be totally "correct" we should implement a hand-off protocol in the watchers -- maybe the watcher can be given an existing reporter?

i'm okay with shipping a "not-correct" version to get others unblocked, and then trying to figure out how to avoid the double-entry in a follow-up PR.